### DISCUSSION TITLE
Feature Request: Ability to Change Docker Command

### DISCUSSION THREAD

**Discussion by nahtnam** [coollabsio/coolify#1689] [Feature Requests]
I would love to be able to set a custom command when deploying a docker image

Current workaround: Create a docker-compose deployment with one service with the custom command

**Comment by AlejandroAkbal** [comment]
I think it's now available, if so, please close this conversation

**Reply by nahtnam** [reply]
I don't think so? I only see these options: https://coolify.io/docs/custom-docker-options

**Comment by andrasbacsai** [comment]
why do you want to change it? can you please give me an example use-case?

**Reply by Wulfheart** [reply]
Not OP but I think it is about e.g. using podman instead of docker. The cli interfaces should be more or less the same but I don't know about the API.

**Reply by nahtnam** [reply]
For example, `glitchtip` requires a webserver and worker. Because custom commands for docker files are not supported, I have to run it like this:

```
FROM glitchtip/glitchtip:latest

CMD ./bin/run-celery-with-beat.sh
```

Not the biggest deal but it is common to have to do this sometimes. It is also possible to do via docker compose, but I prefer to have a different resource for each running task instead of docker compose

**Reply by jsmits** [reply]
I have a similar use-case, also with a worker using the same image, but the worker is started with a different command.

**Reply by nahtnam** [reply]
Another use case would be starting the docker container using https://infisical.com/

**Reply by shintaii** [reply]
Why i would like to change it? Because i do not want to build a seperate docker image, for an application, that only has 1 different argument in the startup. 

So for example i have a container, running an app, that runs in multiple modes. 
In the entrypoint you can make it so that it exec's all commands passed to the start. I.e.:

docker run IMAGE queue1
docker run IMAGE queue2
docker run IMAGE queue2 --debug

**Reply by branch14** [reply]
:+1: I want that feature to run `savonet/liquidsoap`, which is build in a way that it takes the script to run as an argument.

Being able to overwrite the entrypoint via custom docker options would be workaround though.

**Reply by ThomasAunvik** [reply]
@andrasbacsai another case, default for example in for keycloak, they use run arguments to set the runtime.

https://www.keycloak.org/server/containers#_starting_the_optimized_keycloak_container_image

```
podman|docker run --name mykeycloak -p 8443:8443 -p 9000:9000 \
        -e KC_BOOTSTRAP_ADMIN_USERNAME=admin -e KC_BOOTSTRAP_ADMIN_PASSWORD=change_me \
        mykeycloak \
        start --optimized --hostname=localhost
```

And here is how the Dockerfile would look:

https://www.keycloak.org/server/containers#_writing_your_optimized_keycloak_containerfile

```Dockerfile
FROM quay.io/keycloak/keycloak:latest as builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

# Configure a database vendor
ENV KC_DB=postgres

WORKDIR /opt/keycloak
# for demonstration purposes only, please make sure to use proper certificates in production instead
RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:latest
COPY --from=builder /opt/keycloak/ /opt/keycloak/

# change these values to point to a running postgres instance
ENV KC_DB=postgres
ENV KC_DB_URL=<DBURL>
ENV KC_DB_USERNAME=<DBUSERNAME>
ENV KC_DB_PASSWORD=<DBPASSWORD>
ENV KC_HOSTNAME=localhost
ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
```

**Comment by markitosgv** [comment]
same here, workers and web on same image, so I need to use different CMD

**Comment by KMarshland** [comment]
One other workaround I've found is using environment variables to change the CMD dynamically. For my case (a rust app with a server and a worker), I have the CMD set to:
```
CMD ./target/release/${BINARY_NAME:-server}
```

For the worker, I can set the BINARY_NAME environment variable to worker

**Reply by danielehrhardt** [reply]
Cannot add APP_Origin and API_HOST: https://docs.growthbook.io/self-host/env#domains-and-ports

**Comment by zastrixarundell** [comment]
An example of such an application would be a rails application. You start the server with `bundle exec rails s` but you can also start async job processing by running `bundle exec sidekiq`. 

I have never seen a custom Dockerfile for sidekiq.

My company uses EasyPanel which has the option to change the docker command:

![image](https://github.com/user-attachments/assets/85b31866-ae24-48ac-941a-ed8e73da6846)

On my personal projects I have an app which can work either as a webcrawler and/or a server (I know that it should've been seperate but oh well). I can specify via an argument whether it should works with the server or not.

**Comment by theDanielJLewis** [comment]
Any solution for this, yet?

In my case, I'm having trouble trying to make MongoDB work as a shard config server. It seems I have to explicitly add the `--configsvr` argument to the `mongod` command, but I don't see any way to do that inside Coolify. It seems like the only option would be to switch to a custom Docker image or compose setup, but then that gets much more complicated and misses a lot of Coolify's features.

**Reply by theDanielJLewis** [reply]
Aside: the problem I'm running into is more of a [Docker problem](https://github.com/docker-library/mongo/issues/717), because the Mongo Docker image isn't working properly when env username and password are passed. Nonetheless, the workaround (until the Docker image is fixed) is to replace the `command`, which Coolify doesn't support.

**Comment by zastrixarundell** [comment]
Afaik, the simplest way to do this for custom containers is to setup a custom entrypoint which is a bash script and set the default variable as the default execute command.

Something like this:

`coolify-fix.sh`:
```bash
#!/bin/bash

/bin/bash -c "${DEPLOY_COMMAND:-mix phx.server}
```

End of `Dockerfile`:
```dockerfile
ENTRYPOINT [ "./coolify-fix.sh" ]
```

I was then able to run the following commands:
```bash
$ docker build . -t my_app
$ docker run -e DEPLOY_COMMAND="echo 'hello there'" my_app
hello there
```

Of course, this is a workaround and coolify REALLY needs the ability to change the command. This works on an app I made, sure, but won't work if I were to pull an image of another app of which I'm not the maintainer of.

**Reply by twilwa** [reply]
This is currently the situation i'm in -- my dockerfile looks like the following:

```
# Use a specific Node.js version for better reproducibility
FROM node:23.3.0-slim AS builder

# Install pnpm globally and necessary build tools
RUN npm install -g pnpm@9.4.0 && \
    apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y \
        git \
        python3 \
        python3-pip \
        curl \
        node-gyp \
        ffmpeg \
        libtool-bin \
        autoconf \
        automake \
        libopus-dev \
        make \
        g++ \
        build-essential \
        libcairo2-dev \
        libjpeg-dev \
        libpango1.0-dev \
        libgif-dev \
        openssl \
        libssl-dev && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Set Python 3 as the default python
RUN ln -sf /usr/bin/python3 /usr/bin/python

# Set the working directory
WORKDIR /app

# Copy application code
COPY . .

# Install dependencies
RUN pnpm install --no-frozen-lockfile

# Build the project
RUN pnpm run build && pnpm prune --prod

# Final runtime image
FROM node:23.3.0-slim

# Install runtime dependencies
RUN npm install -g pnpm@9.4.0 && \
    apt-get update && \
    apt-get install -y \
        git \
        python3 \
        ffmpeg && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Set the working directory
WORKDIR /app

# Copy built artifacts and production dependencies from the builder stage
COPY --from=builder /app/package.json ./
COPY --from=builder /app/pnpm-workspace.yaml ./
COPY --from=builder /app/eslint.config.mjs ./
COPY --from=builder /app/.eslintrc.json ./
COPY --from=builder /app/.npmrc ./
COPY --from=builder /app/turbo.json ./
COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/agent ./agent
COPY --from=builder /app/client ./client
COPY --from=builder /app/lerna.json ./
COPY --from=builder /app/packages ./packages
COPY --from=builder /app/scripts ./scripts
COPY --from=builder /app/characters ./characters

# Expose necessary ports
EXPOSE 3000 5173

# Command to start the application
CMD ["sh", "-c", "pnpm start & pnpm start:client"]
```
I need to run the start command with args pointing to custom application configuation files, but coolify doesn't seem to support this. As it stands it seems I'll have to fork the repo and create a branch for the coolify script; ideally I'd prefer not to do this/I imagine there may be situations where others are working with closed source code that a client doesn't want leaving the gh org. Seems like fairly basic functionality to include, and could simply copy-paste the interface for overwriting docker-composes to include dockerfiles.

**Comment by theDanielJLewis** [comment]
Another use case: [making Dragonfly work with BullMQ](https://www.dragonflydb.io/docs/integrations/bullmq).

This requires adding a flag on the Dragonfly start command. But adding such a flag is not possible to do with Coolify unless I go the complicated route of making and managing my own Docker Compose.

**Reply by boredland** [reply]
uh, exactly my use-case. Very sad that I can't run with a specific command or at least "eject" the existing config into a custom docker deployment.

**Reply by adamhancock** [reply]
[Have you tried using environment variables?](https://github.com/taskforcesh/bullmq-proxy/blob/1cb5156622074ba75028ffa871f4aad2e5d18a42/docker-compose-dfly.yaml#L18-L19)

**Reply by boredland** [reply]
Oh, sorry, yes. That is actually how I solved that for us!

**Reply by agulencina96** [reply]
> [Have you tried using environment variables?](https://github.com/taskforcesh/bullmq-proxy/blob/1cb5156622074ba75028ffa871f4aad2e5d18a42/docker-compose-dfly.yaml#L18-L19)

Thank you so much!

**Reply by mjy78** [reply]
Is there a way to add environment variables for setting a snapshot schedule?

I can't seem to find any way to do this other than by modifying the Dragonfly command to include --snapshot_cron parameter (which I can't seem to do with the Coolify Dragonfly template). 

Another approach I thought might be possible would be to add a "Scheduled Task" to run the redis-cli SAVE command regularly in the container, but it seems the Coolify Dragonfly template doesn't expose "Scheduled Task" in the UI either.

**Comment by xboxfly15** [comment]
Another example is https://docs.victoriametrics.com/victorialogs/quickstart/#docker-image
Need custom command to define options like retention, http port etc.

**Reply by djsisson** [reply]
Just paste in below as a new compose service

```yaml
services:
  victorialogs:
    image: victoriametrics/victoria-logs:v1.31.0
    ports:
      - "9428:9428"
    volumes:
      - victoria-logs-data:/victoria-logs-data
    command: >
      -storageDataPath=/victoria-logs-data
      -retentionPeriod=1m
      -httpListenAddr=:9428
    healthcheck:
      test: ["CMD", "wget", "-q", "--spider", "http://localhost:9428"]
      interval: 30s
      timeout: 10s
      retries: 5
```

of use part of their compose
https://github.com/VictoriaMetrics/VictoriaLogs/blob/master/deployment/docker/victorialogs/compose-base.yml

**Comment by adiologydev** [comment]
I've since made a PR to resolve this: https://github.com/coollabsio/coolify/pull/7097

**Reply by adiologydev** [reply]
It's live in 450.

**Comment by theDanielJLewis** [comment]
This issue should _not_ be closed because it has _not_ been resolved. v450 only added the ability to change `--entrypoint`. It didn't add the option to change the run command for the Docker container.

**Reply by adiologydev** [reply]
Could you please share what you mean by run command? Because when we look at platforms like Railway who offer a Run Command option, they use entrypoint for that container which is now available and works.

If it's something else, I'd love to know.

**Reply by theDanielJLewis** [reply]
@adiologydev Please just search this page for my username. I already posted a couple of specific cases: Dragonfly and MongoDB.

**Reply by ThomasAunvik** [reply]
@adiologydev in the compose spec, `entrypoint` overrides the docker image spec.

> Override the default entrypoint declared by the container image, for example 'ENTRYPOINT' in Dockerfile

https://raw.githubusercontent.com/compose-spec/compose-spec/master/schema/compose-spec.json

You would use `command` to add additional arguments if there is an `ENTRYPOINT` assigned.

https://github.com/compose-spec/compose-spec/blob/main/spec.md#command

https://www.docker.com/blog/docker-best-practices-choosing-between-run-cmd-and-entrypoint/

**Reply by adiologydev** [reply]
Okay I think I understand what you mean now, I wasn't sure if entrypoint would completely overwrite the default CMD arguments or not.

I will see if I can make a PR to support this option as well.
